### PR TITLE
Shared refs

### PR DESCRIPTION
**Pull request by ryzhyk**
Details in commit messages.

### CODE REVIEW

File: test/datalog_tests/internment_test.dl [hunk 1]
```diff
@@ -20,3 +20,29 @@ OInternedString(ival(s), s) :-
...
+
+input relation IStruct[Intern<IStruct>]
+
+output relation Projections(inp: Intern<IStruct>, p: string)
+
+Projections(i, "x=${i.x}") :-
+    IStruct[i].
+Projections(i, "t.0=${i.t.0}") :-
+    i in &IStruct().
+Projections(i, "t.1=${i.t.1}") :-
```
Comment (mbudiu-vmw) [inline test/datalog_tests/internment_test.dl:41]: I suspect you don't need the d here

File: lib/intern.dl [hunk 2]
```diff
@@ -14,6 +14,7 @@
  * There is simply no way to obtain an interned object of a different type.
  */
 #[size=4]
+#[shared_ref]
```
Comment (mbudiu-vmw) [inline lib/intern.dl:17]: I don't really grasp what the user should know about this annotation.
I don't really understand what it does; I suspect it's really tied to the Rust implementation.

**Comment by mbudiu-vmw** [comment]
So is the point of this to allow interned and references types to be essentially the same?

File: src/Language/DifferentialDatalog/Attribute.hs [hunk 3]
```diff
@@ -70,6 +71,11 @@ typedefValidateAttr d TypeDef{..} attr = do
             let t = fromJust tdefType
             check d (isStruct d t) (pos attr)
                 $ "'custom_serde' attribute cannot be applied to type aliases."
+         "shared_ref" -> do
+            check d (isNothing tdefType) (pos attr)
+                $ "'sharef_ref' attribute is only applicable to extern types."
+            check d (length tdefArgs == 1) (pos attr)
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/Language/DifferentialDatalog/Attribute.hs:77]: Sorry, I'm not sure what you are saying.
We are checking that the annotated type has exactly one type argument, e.g., `Ref<'T>` is ok, but `Ref<'T1, 'T2>` isn't.

File: lib/intern.dl (hunk 2 above)
Comment (ryzhyk) [reply lib/intern.dl:17]: It is. I don't expect DDlog users to use this annotation, but it allows us to introduce new smart pointers in the future without changing the compiler.

File: src/Language/DifferentialDatalog/Attribute.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply src/Language/DifferentialDatalog/Attribute.hs:77]: I guess I am asking for a message more like: 'types annotated with shared_ref must have exactly one type argument'

**Comment by ryzhyk** [comment]
> So is the point of this to allow interned and references types to be essentially the same?

They can be used in similar ways, but as a user you still have to know which one is better for your application.

File: src/Language/DifferentialDatalog/Attribute.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/Language/DifferentialDatalog/Attribute.hs:77]: That makes sense, thanks!

**Pull request merged**